webapp/graphite/tags/localdatabase.py

```diff
@@ -0,0 +1,215 @@
+        sql += ' LEFT JOIN tags_tagvalue AS v' + s + ' ON v' + s + '.id=st' + s + '.value_id AND v' + s + '.value REGEXP %s'
+        params.append(value)
+
+        where.append('v' + s + '.id IS NULL')
+
+    if where:
+      sql += ' WHERE ' + ' AND '.join(where)
+      params.extend(whereparams)
+
+    return sql, params
+
+  def find_series(self, tags):
+    sql, params = self.find_series_query(tags)
+
+    with connection.cursor() as cursor:
+      cursor.execute(sql, params)
+
+      return [row[0] for row in cursor.fetchall()]
+
+  def get_series(self, path):
+    with connection.cursor() as cursor:
+      sql = 'SELECT s.id, t.tag, v.value'
+      sql += ' FROM tags_series AS s'
+      sql += ' JOIN tags_seriestag AS st ON st.series_id=s.id'
+      sql += ' JOIN tags_tag AS t ON t.id=st.tag_id'
+      sql += ' JOIN tags_tagvalue AS v ON v.id=st.value_id'
+      sql += ' WHERE s.path=%s'
+      params = [path]
+      cursor.execute(sql, params)
+
+      tags = {tag: value for (id, tag, value) in cursor.fetchall()}
+
+      if not tags:
+        return None
+
+      return TaggedSeries(tags['name'], tags, id=id)
+
+  def list_tags(self):
+    with connection.cursor() as cursor:
+      sql = 'SELECT t.id, t.tag'
+      sql += ' FROM tags_tag AS t'
+      params = []
+      cursor.execute(sql, params)
+
+      return [{'id': id, 'tag': tag} for (id, tag) in cursor.fetchall()]
+
+  def get_tag(self, tag):
+    with connection.cursor() as cursor:
+      sql = 'SELECT t.id, t.tag'
+      sql += ' FROM tags_tag AS t'
+      sql += ' WHERE t.tag=%s'
+      params = [tag]
+      cursor.execute(sql, params)
+
+      row = cursor.fetchone()
+
+    if not row:
+      return None
+
+    (id, tag) = row
+
+    return {
+      'id': id,
+      'tag': tag,
+      'values': self.list_values(tag),
+    }
+
+  def list_values(self, tag):
+    with connection.cursor() as cursor:
+      sql = 'SELECT v.id, v.value, COUNT(st.id)'
+      sql += ' FROM tags_tagvalue AS v'
+      sql += ' JOIN tags_seriestag AS st ON st.value_id=v.id'
+      sql += ' JOIN tags_tag AS t ON t.id=st.tag_id'
+      sql += ' WHERE t.tag=%s'
+      sql += ' GROUP BY v.id, v.value'
+      params = [tag]
+      cursor.execute(sql, params)
+
+      return [{'id': id, 'value': value, 'count': count} for (id, value, count) in cursor.fetchall()]
+
+  def _insert_ignore(self, table, cols, data):
+    if connection.vendor == 'mysql':
+      sql = 'INSERT IGNORE INTO '
+    elif connection.vendor == 'sqlite':
+      sql = 'INSERT OR IGNORE INTO '
+    elif connection.vendor == 'postgresql':
+      sql = 'INSERT IGNORE INTO '
+    else:
+      raise Exception('Unsupported database vendor ' + connection.vendor)
+
+    sql += table + ' (' + ','.join(cols) + ') VALUES ' + ', '.join(['(' + ', '.join(['%s'] * len(cols)) + ')'] * len(data))
+    params = []
+    for row in data:
```

Comment:
Nope, `params` is a list containing all the values of each item in `data`, it's using `extend` not `append`.